File: solutions/python/25.py
def execute(code: str) -> int:
    value = 0
    index = 0

    while index < len(code):
        instruction = code[index]
        if instruction == "+":
            value += 1
        elif instruction == "-":
            value -= 1
        elif instruction == "[" and value == 0:
            index = code.find("]", index)
        elif instruction == "]" and value != 0:
            index = code.rfind("[", 0, index)
        elif instruction == "{" and value == 0:
            index = code.find("}", index)

        index += 1

    return value


execute1 = execute("+++")  # 3
print(execute1)

execute2 = execute("+--")  # -1
print(execute2)

execute3 = execute(">+++[-]")  # 0
print(execute3)

execute4 = execute(">>>+{++}")  # 3
print(execute4)

execute5 = execute("+{[-]+}+")  # 2
print(execute5)

execute6 = execute("{+}{+}{+}")  # 0
print(execute6)

execute7 = execute("------[+]++")  # 2
print(execute7)

execute8 = execute("-[++{-}]+{++++}")  # 5
print(execute8)

"""
  Score: ★★★★★★ (3 bonus stars for solving it today!)
  Points: 250
  ops: 4154
"""

File: solutions/ts/25.ts
function execute(code: string): number {
  let v = 0,
    i = 0;

  const actions: { [key: string]: () => void } = {
    "+": () => v++,
    "-": () => v--,
    "[": () => v === 0 && (i = code.indexOf("]", i)),
    "]": () => v !== 0 && (i = code.lastIndexOf("[", i)),
    "{": () => v === 0 && (i = code.indexOf("}", i)),
  };

  while (i < code.length) {
    actions[code[i]]?.();
    i++;
  }

  return v;
}

const execute1 = execute("+++"); // 3
console.log(execute1);

const execute2 = execute("+--"); // -1
console.log(execute2);

const execute3 = execute(">+++[-]"); // 0
console.log(execute3);

const execute4 = execute(">>>+{++}"); // 3
console.log(execute4);

const execute5 = execute("+{[-]+}+"); // 2
console.log(execute5);

const execute6 = execute("{+}{+}{+}"); // 0
console.log(execute6);

const execute7 = execute("------[+]++"); // 2
console.log(execute7);

const execute8 = execute("-[++{-}]+{++++}"); // 5
console.log(execute8);

/* 
  Score: ★★★★★★★ (3 bonus stars for solving it today!)
  Points: 250
  ops: 3946
*/
